core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -54,5 +77,228 @@ func (ft *GCSFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
 func (ft *GCSFileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
 	ft.logger.Debug("GCSFileTransfer: Download: downloading file", "path", task.Path, "ref", task.Reference)
 
-	return fmt.Errorf("not implemented yet")
+	reference := task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, objectName, err := parseReference(reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+	bucket := ft.client.Bucket(bucketName)
+
+	// If not using checksum, we set the digest to the reference
+	if task.Digest == reference {
+		// Get all of the objects under the reference
+		var objectNames []string
+		query := &storage.Query{Prefix: objectName}
+		it := bucket.Objects(ft.ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return formatDownloadError("error iterating through objects", err)
+			}
+			objectNames = append(objectNames, objAttrs.Name)
+		}
+
+		// Try to download all objects under the reference
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object, _, err := ft.getObjectAndAttrs(bucket, task, objName)
+				if err != nil {
+					if errors.Is(err, ErrObjectIsDirectory) {
+						ft.logger.Debug(
+							"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+							"reference", reference,
+							"object", objName,
+						)
+						return nil
+					}
+					return formatDownloadError(fmt.Sprintf("error getting object %s", objName), err)
+				}
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
```

Comment:
I would expect the file to be downloaded to `task.Path`, based on the field's name and documentation.

Reply:
My point is that the name and documentation need to be updated, or else this code is surprising. As is, `PathOrPrefix` is a better name, but it would be even simpler if the field were to *always* store a prefix and the final name were *always* constructed by concatenating the prefix and the name; in that case, you could call the field `PathPrefix`. Is that possible?

If a `PathPrefix` field complicates caller logic, then you can use `PathOrPrefix` with a doc comment explaining how to determine which it is.

Reply:
gotcha, we could use the same `getDownloadFilePath` function to construct the download path. in that case, the prefix and the name would be the same so the extension would just be an empty string (because we remove anything in common between the two), giving us the correct path anyway